### PR TITLE
Clean up ALBs using `spec.ingressClassName` and ALB security groups

### PR DESCRIPTION
**Pull request by aaroniscode**
### Description

ALBs created using `ingressClassName` in the Ingress spec are not cleaned up today. The code was only looking at the deprecated `kubernetes.io/ingress.class` annotation.

The code now checks if `ingressClassName` is set and if not, falls back to checking the annotation.

This is my first PR for eksctl so please guide me on required documentation. I didn't see any tests and wasn't sure if it was necessary. Let me know. Thanks.

Edit: added a second commit that fixes deletion of ALB security groups. It doesn't delete them, it just waits for the AWS LB Controller to delete them before deleting the node group.

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [X] Manually tested
- [X] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Review by Himangini — COMMENTED** [review]
@aaroniscode Thanks for the contribution. Can you perhaps write up how you manually tested the changes?

**Comment by aaroniscode** [comment]
sure @Himangini

Testing was straightforward. Created a cluster using eksctl

`eksctl create cluster -f alb.yaml`

alb.yaml below:
```
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: alb
  region: us-west-2
  version: "1.25"

addons:
- name: vpc-cni

cloudWatch:
  clusterLogging:
    enableTypes: ["*"]

iam:
  withOIDC: true

managedNodeGroups:
- name: main
  amiFamily: AmazonLinux2
  iam:
    attachPolicyARNs:
    - arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy
    - arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly
    - arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore
  instanceType: t3.large
  minSize: 0
  desiredCapacity: 2
  maxSize: 10
  privateNetworking: true
  spot: false
```

Installed AWS LB Controller using instructions here: https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html.

Deploy the example game 2048 application
```
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: game-2048
  name: deployment-2048
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: app-2048
  replicas: 1
  template:
    metadata:
      labels:
        app.kubernetes.io/name: app-2048
    spec:
      containers:
      - image: public.ecr.aws/l6m2t8p7/docker-2048:latest
        imagePullPolicy: Always
        name: app-2048
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  namespace: game-2048
  name: service-2048
spec:
  ports:
    - port: 80
      targetPort: 80
      protocol: TCP
  type: ClusterIP
  selector:
    app.kubernetes.io/name: app-2048
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  namespace: game-2048
  name: ingress-2048
  annotations:
    alb.ingress.kubernetes.io/scheme: internet-facing
    alb.ingress.kubernetes.io/target-type: ip
spec:
  ingressClassName: alb
  rules:
    - http:
        paths:
        - path: /
          pathType: Prefix
          backend:
            service:
              name: service-2048
              port:
                number: 80
 ```

Then delete the cluster `eksctl delete cluster alb`

File: pkg/elb/cleanup.go [hunk 1]
```diff
@@ -364,32 +377,50 @@ func describeSecurityGroupsByID(ctx context.Context, ec2API awsapi.EC2, groupIDs
...
 	}
 	return false
 }
 
 func getSecurityGroupsOwnedByLoadBalancer(ctx context.Context, ec2API awsapi.EC2, elbAPI DescribeLoadBalancersAPI,
-	clusterName string, loadBalancerName string, loadBalancerKind loadBalancerKind) (map[string]struct{}, error) {
-	if loadBalancerKind == network {
+	elbv2API DescribeLoadBalancersAPIV2, clusterName string, loadBalancerName string, loadBalancerKind loadBalancerKind) (map[string]struct{}, error) {
+
+	groupIds := []string{}
```
Comment (cPu1) [inline pkg/elb/cleanup.go:392]: Minor nit:
```suggestion
	var groupIDs []string
```

File: pkg/elb/cleanup.go [hunk 2]
```diff
@@ -459,6 +490,30 @@ func elbExists(ctx context.Context, elbAPI DescribeLoadBalancersAPI, elbv2API De
...
 }
 
+func describeApplicationLoadBalancer(ctx context.Context, elbv2API DescribeLoadBalancersAPIV2,
+	name string) (*elbv2types.LoadBalancer, error) {
+
+	response, err := elbv2API.DescribeLoadBalancers(ctx, &elasticloadbalancingv2.DescribeLoadBalancersInput{
+		Names: []string{name},
+	})
+	if err != nil {
+		if isELBNotFoundErr(err) {
```
Comment (cPu1) [inline pkg/elb/cleanup.go:500]: A not-found error returned from `elbv2API.DescribeLoadBalancers` will not pass `isELBNotFoundErr`. The error returned would be of type `elasticloadbalancingv2types.LoadBalancerNotFoundException` which is not handled by `isELBNotFoundErr`.

**Review by cPu1 — COMMENTED** [review]
I have left some suggestions but otherwise it LGTM 🙂.

File: pkg/elb/cleanup.go [hunk 3]
```diff
@@ -193,37 +199,44 @@ func getServiceLoadBalancer(ctx context.Context, ec2API awsapi.EC2, elbAPI Descr
...
 	if len(hostNameParts[0]) == 0 {
 		logger.Debug("%s is ALB Ingress, but probably not provisioned or something other unexpected, skip", metadata.Name)
-		return nil
+		return nil, nil
 	}
 	name := strings.TrimPrefix(hostNameParts[0], "internal-") // Trim 'internal-' prefix for ALB DNS name which is not a part of name.
-	if len(name) > 31 {
-		name = name[:31]
+	if len(name) > 32 {
+		name = name[:32]
```
[the remark this answers is not in the document]
Comment (aaroniscode) [reply pkg/elb/cleanup.go:227]: 🙌

File: pkg/elb/cleanup.go [hunk 4]
```diff
@@ -193,37 +199,44 @@ func getServiceLoadBalancer(ctx context.Context, ec2API awsapi.EC2, elbAPI Descr
...
+	}
+	ctx, cleanup := context.WithTimeout(ctx, 30*time.Second)
+	securityGroupIDs, err := getSecurityGroupsOwnedByLoadBalancer(ctx, ec2API, elbAPI, elbv2API, clusterName, name, application)
+	cleanup()
```
[the remark this answers is not in the document]
Comment (aaroniscode) [reply pkg/elb/cleanup.go]: done

File: pkg/elb/cleanup.go [hunk 5]
```diff
@@ -119,8 +120,13 @@ func Cleanup(ctx context.Context, ec2API awsapi.EC2, elbAPI DescribeLoadBalancer
 	}
 	// For k8s Kind Ingress
 	for _, i := range ingresses {
-		lb := getIngressLoadBalancer(i)
 		ingressMetadata := i.GetMetadata()
+
+		lb, err := getIngressLoadBalancer(ctx, ec2API, elbAPI, elbv2API, clusterConfig.Metadata.Name, i)
+		if err != nil {
+			return fmt.Errorf("cannot obtain information for ALB from Ingress %s/%s: %s",
```
[the remark this answers is not in the document]
Comment (aaroniscode) [reply pkg/elb/cleanup.go]: done

File: pkg/elb/cleanup.go (hunk 2 above)
Comment (aaroniscode) [reply pkg/elb/cleanup.go]: thanks for catching this ... I added a new `isELBv2NotFoundErr` function that looks for the correct exception in the right package.

**Comment by aaroniscode** [comment]
@cPu1 thanks for the code review and the suggestions!  I applied all of your suggestions in a new commit.

**Pull request merged**